File: example/src/main/java/com/github/vivchar/example/pages/simple/InputsFragment.java
package com.github.vivchar.example.pages.simple;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.github.vivchar.example.BaseScreenFragment;
import com.github.vivchar.example.R;
import com.github.vivchar.example.widgets.BetweenSpacesItemDecoration;
import com.github.vivchar.example.widgets.MyAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.ViewFinder;
import com.github.vivchar.rendererrecyclerviewadapter.ViewHolder;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer;
import com.github.vivchar.rendererrecyclerviewadapter.ViewState;
import com.github.vivchar.rendererrecyclerviewadapter.ViewStateProvider;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

public class InputsFragment extends BaseScreenFragment {

	public static final String TAG = ViewRendererFragment.class.getSimpleName();

	private final YourDataProvider mYourDataProvider = new YourDataProvider();

	@Nullable
	@Override
	public View onCreateView(final LayoutInflater inflater,
	                         @Nullable final ViewGroup container,
	                         @Nullable final Bundle savedInstanceState) {

		final View view = inflater.inflate(R.layout.fragment_list, container, false);

		final RendererRecyclerViewAdapter adapter = new MyAdapter();

		adapter.registerRenderer(new ViewRenderer<>(
				R.layout.item_input,
				InputViewModel.class,
				(model, finder, payloads) -> finder.setText(R.id.text, model.getValue()),
				getViewStateProvider()
		));

		adapter.setItems(mYourDataProvider.getInputsModels());

		final RecyclerView recyclerView = view.findViewById(R.id.recycler_view);
		recyclerView.setAdapter(adapter);
		recyclerView.addItemDecoration(new BetweenSpacesItemDecoration(10, 10));

		return view;
	}

	@NonNull
	private ViewStateProvider<InputViewModel, ViewHolder<ViewFinder>> getViewStateProvider() {
		return new ViewStateProvider<InputViewModel, ViewHolder<ViewFinder>>() {
			@Override
			public ViewState createViewState() {
				return new ViewState<ViewHolder<ViewFinder>>() {
					private String savedValue; /* here will be saved a value for each item by createViewStateID() */

					@Override
					public void clear(@NonNull final ViewHolder<ViewFinder> holder) {
						holder.getViewFinder().setText(R.id.text, "");
					}

					@Override
					public void save(@NonNull final ViewHolder<ViewFinder> holder) {
						savedValue = holder.getViewFinder().<EditText>find(R.id.text).getText().toString();
					}

					@Override
					public void restore(@NonNull final ViewHolder<ViewFinder> holder) {
						holder.getViewFinder().setText(R.id.text, savedValue);
					}
				};
			}

			@Override
			public int createViewStateID(@NonNull final InputViewModel model) {
				return model.getID();
			}
		};
	}

	public static class InputViewModel implements ViewModel {

		private final int mID;
		@NonNull
		private final String mValue;

		public InputViewModel(final int ID, @NonNull final String value) {
			mID = ID;
			mValue = value;
		}

		private int getID() {
			return mID;
		}

		@NonNull
		public String getValue() {
			return mValue;
		}
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/InputsFragment.kt
package com.github.vivchar.example.pages.simple

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.github.vivchar.example.R
import com.github.vivchar.example.base.BaseFragment
import com.github.vivchar.example.databinding.FragmentListBinding
import com.github.vivchar.example.widgets.BetweenSpacesItemDecoration
import com.github.vivchar.example.widgets.MyAdapter
import com.github.vivchar.rendererrecyclerviewadapter.*
import kotlinx.coroutines.launch

class InputsFragment : BaseFragment<FragmentListBinding>() {

	private val viewModel: InputsViewModelVM by viewModels()
	private val adapter = MyAdapter()

	override fun createBinding(inflater: LayoutInflater, container: ViewGroup?) =
		FragmentListBinding.inflate(inflater, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		adapter.registerRenderer(
			ViewRenderer(
				R.layout.item_input,
				InputViewModel::class.java,
				{ model, finder, _ -> finder.setText(R.id.text, model.value) },
				viewStateProvider
			)
		)

		binding.recyclerView.adapter = adapter
		binding.recyclerView.addItemDecoration(BetweenSpacesItemDecoration(10, 10))

		viewLifecycleOwner.lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				viewModel.state.collect { state ->
					adapter.setItems(state.items)
				}
			}
		}
	}

	private val viewStateProvider: ViewStateProvider<InputViewModel, ViewHolder<ViewFinder>>
		get() = object : ViewStateProvider<InputViewModel, ViewHolder<ViewFinder>> {
			override fun createViewState(): ViewState<*> {
				return object : ViewState<ViewHolder<ViewFinder>> {
					private var savedValue: String? = null

					override fun clear(holder: ViewHolder<ViewFinder>) {
						holder.viewFinder.setText(R.id.text, "")
					}

					override fun save(holder: ViewHolder<ViewFinder>) {
						savedValue = holder.viewFinder.find<EditText>(R.id.text).text.toString()
					}

					override fun restore(holder: ViewHolder<ViewFinder>) {
						holder.viewFinder.setText(R.id.text, savedValue)
					}
				}
			}

			override fun createViewStateID(model: InputViewModel) = model.id
		}

	data class InputViewModel(val id: Int, val value: String) : ViewModel
}
